Source/Scene/Cesium3DTileBatchTable.js

```diff
@@ -341,32 +453,263 @@ define([
+        var componentCount = binaryProperty.componentCount;
+        if (componentCount === 1) {
+            typedArray[index] = value;
+        } else {
+            binaryProperty.type.pack(value, typedArray, index * componentCount);
+        }
+    }
+
+    var scratchStack = [];
+    var marker = 0;
+    function traverseHierarchyTree(hierarchy, instanceIndex, endConditionCallback) {
+        var parentCounts = hierarchy.parentCounts;
+        var parentIds = hierarchy.parentIds;
+        var parentIndexes = hierarchy.parentIndexes;
+
+        // Ignore instances that have already been visited. This occurs in diamond inheritance situations.
+        // Use a marker value to indicate that an instance has been visited, which increments with each run.
+        // This is more efficient than clearing the visited array every time.
+        var visited = hierarchy.visited;
+        var visitedMarker = ++marker;
+
+        var stack = scratchStack;
+        stack.length = 0;
+        stack.push(instanceIndex);
+
+        while (stack.length > 0) {
+            instanceIndex = stack.pop();
+            if (visited[instanceIndex] === visitedMarker) {
+                // This instance has already been visited, stop traversal
+                continue;
+            }
+            visited[instanceIndex] = visitedMarker;
+            var result = endConditionCallback(hierarchy, instanceIndex, stack);
+            if (defined(result)) {
+                // The end condition was met, stop the traversal and return the result
+                return result;
+            }
+            var parentCount = defined(parentCounts) ? parentCounts[instanceIndex] : 1;
+            var parentIndex = defined(parentCounts) ? parentIndexes[instanceIndex] : instanceIndex;
+            for (var i = 0; i < parentCount; ++i) {
+                var parentId = parentIds[parentIndex + i];
+                // Stop the traversal when the instance has no parent (its parentId equals itself)
+                // else add the parent to the stack to continue the traversal.
+                if (parentId !== instanceIndex) {
+                    stack.push(parentId);
+                }
+            }
+        }
+    }
+
+    function traverseHierarchyLinear(hierarchy, instanceIndex, endConditionCallback) {
+        while (true) {
+            var result = endConditionCallback(hierarchy, instanceIndex);
+            if (defined(result)) {
+                // The end condition was met, stop the traversal and return the result
+                return result;
+            }
+            var parentId = hierarchy.parentIds[instanceIndex];
+            if (parentId === instanceIndex) {
+                // Stop the traversal when the instance has no parent (its parentId equals itself)
+                break;
+            }
+            instanceIndex = parentId;
+        }
+    }
+
+    function traverseHierarchy(hierarchy, instanceIndex, endConditionCallback) {
```

Comment:
I'm pretty confident that the other functions test it sufficiently, but like your comment below there could be more test cases.